pkg/natdiscovery/request_send.go

```diff
@@ -0,0 +1,152 @@
+}
+
+func (nd *natDiscovery) sendCheckRequest(remoteNAT *remoteEndpointNAT) error {
+	var errPrivate, errPublic error
+	var reqID uint64
+	if remoteNAT.Endpoint.Spec.PrivateIP != "" {
+		reqID, errPrivate = nd.sendCheckRequestToTargetIP(remoteNAT, remoteNAT.Endpoint.Spec.PrivateIP)
+		if errPrivate == nil {
+			remoteNAT.lastPrivateIPRequestID = reqID
+		}
+	}
+
+	if remoteNAT.Endpoint.Spec.PublicIP != "" {
+		reqID, errPublic = nd.sendCheckRequestToTargetIP(remoteNAT, remoteNAT.Endpoint.Spec.PublicIP)
+		if errPublic == nil {
+			remoteNAT.lastPublicIPRequestID = reqID
+		}
+	}
+
+	if errPrivate != nil && errPublic != nil {
+		return errors.Errorf("error happened trying to discover both public & private IPs of endpoint %q, [%s, %s]",
+			remoteNAT.Endpoint.Spec.CableName, errPublic, errPrivate)
+	}
+
+	if errPrivate != nil {
+		return errors.Wrapf(errPrivate, "error happened trying to NAT-discover private IP of endpoint %q",
+			remoteNAT.Endpoint.Spec.CableName)
+	}
+
+	if errPublic != nil {
+		return errors.Wrapf(errPublic, "error happened trying to NAT-discover public IP of endpoint %q",
+			remoteNAT.Endpoint.Spec.CableName)
+	}
+
+	return nil
+}
+
+func (nd *natDiscovery) sendCheckRequestToTargetIP(remoteNAT *remoteEndpointNAT, targetIP string) (uint64, error) {
+	nd.requestCounter++
+
+	targetPort, err := extractNATDiscoveryPort(remoteNAT.Endpoint)
+
+	if err != nil {
+		return 0, err
+	}
+
+	sourceIP, err := nd.findSrcIP(targetIP)
+	if err != nil {
+		klog.Warningf("unable to determine source IP while preparing NAT discovery request to endpoint %q: %s",
+			remoteNAT.Endpoint.Spec.CableName, err)
+	}
+
+	request := &natproto.SubmarinerNatDiscoveryRequest{
+		RequestNumber: nd.requestCounter,
+		Sender: &natproto.EndpointDetails{
+			EndpointId: nd.localEndpoint.Spec.CableName,
+			ClusterId:  nd.localEndpoint.Spec.ClusterID,
+		},
+		Receiver: &natproto.EndpointDetails{
+			EndpointId: remoteNAT.Endpoint.Spec.CableName,
+			ClusterId:  remoteNAT.Endpoint.Spec.ClusterID,
+		},
+		UsingSrc: &natproto.IPPortPair{
+			IP:   sourceIP,
+			Port: uint32(nd.serverPort),
+		},
+		UsingDst: &natproto.IPPortPair{
+			IP:   targetIP,
+			Port: uint32(targetPort),
+		},
+	}
+
+	msgRequest := &natproto.SubmarinerNatDiscoveryMessage_Request{
+		Request: request,
+	}
+
+	message := natproto.SubmarinerNatDiscoveryMessage{
+		Version: natproto.Version,
+		Message: msgRequest,
+	}
+
+	buf, err := proto.Marshal(&message)
+	if err != nil {
+		return request.RequestNumber, errors.Wrapf(err, "error marshaling request %#v", request)
+	}
+
+	addr := net.UDPAddr{
+		IP:   net.ParseIP(targetIP),
+		Port: targetPort,
+	}
+
+	if length, err := nd.serverUDPWrite(buf, &addr); err != nil {
+		return request.RequestNumber, errors.Wrapf(err, "error sending request packet %#v", request)
+	} else if length != len(buf) {
+		return request.RequestNumber, errors.Errorf("the sent UDP packet was smaller than requested, sent=%d, expected=%d", length,
```

Comment:
As above: are we retrying when only partial data is sent?

Reply:
It will be retried by the controller on the upper level after recheckTime (2 seconds)